### DISCUSSION TITLE
Gallery: Post Your Screenshots / Code Here

### DISCUSSION THREAD

**Discussion by epezent** [epezent/implot#180] [Show and tell]
Post your screenshots or custom code here and tell us about your project!

**Comment by epezent** [comment]
**Custom candlestick plots and tool tips:**

![candletool](https://user-images.githubusercontent.com/29577475/90851160-7b548880-e339-11ea-90b5-27b993421005.gif)

**ImPlot + ImGui Tables API:**

![tables](https://user-images.githubusercontent.com/29577475/90851187-8c04fe80-e339-11ea-8267-e6bfdc105686.gif)

Both of these have been added to `implot_demo.cpp`

**Comment by PeterJohnson** [comment]
Robot simulation GUI for FIRST robotics with multiple plots and drag and drop everything.


![drag and drop resized](https://user-images.githubusercontent.com/297126/91929143-9a84db80-ec92-11ea-80e1-3eea3ff8b890.gif)


https://github.com/wpilibsuite/allwpilib/pull/2653

**Comment by epezent** [comment]
@PeterJohnson , I'm curious how you got the linked axes working and if it is bidirectional. This is a feature I've wanted to implement.

**Comment by hoffstadt** [comment]
Some things DearPyGui users have made:

![fa08r61ae2i51](https://user-images.githubusercontent.com/39973752/92046481-3fadbb80-ed48-11ea-9b3b-9cc5ea6e3a4e.png)

![114vv9n3ask51](https://user-images.githubusercontent.com/39973752/92046084-c910be00-ed47-11ea-9a33-7a761eb7d356.jpg)

![91245835-0c4daa00-e714-11ea-8b3f-6ac01d6dbeb3](https://user-images.githubusercontent.com/39973752/92046379-ffe6d400-ed47-11ea-9a19-ab672fa9c26f.gif)
![89934768-b686e700-dbd6-11ea-9433-33c123f92a01](https://user-images.githubusercontent.com/39973752/92046423-1f7dfc80-ed48-11ea-9e91-5dc615e22bfd.png)

**Comment by PeterJohnson** [comment]
@epezent It is not bidirectional.  I just use SetNextPlotLimitsX() to force the X axis on the second plot to follow the first.  Bidirectional is a good idea.  I think how I would tackle that is to do the SetNextPlotLimitsX() on the other plots only when one of them is actively being dragged (there would be a 1-frame delay to updating the first plot when you dragged the second plot, but that's hardly noticeable).  However, to do this, I would need a method to detect when a plot is being dragged by the user.  Is that exposed in implot_internal.h?

**Comment by marcizhu** [comment]
Not sure if this one counts, but since I will be using this library to make an application for USB oscilloscopes, I've been doing some experiments where I plotted some (generated) signals into heatmaps in order to visually differentiate the random noise from the real signal. Here's the result:

<img width="1392" alt="img1" src="https://user-images.githubusercontent.com/6199781/98874475-9f2bf400-247a-11eb-9f6a-850c1db2fefc.png">
<img width="1392" alt="img2" src="https://user-images.githubusercontent.com/6199781/98874504-aa7f1f80-247a-11eb-8fda-e2555728c72d.png">
<img width="1392" alt="img3" src="https://user-images.githubusercontent.com/6199781/98874557-c682c100-247a-11eb-801a-82340750bb11.png">
<img width="1392" alt="img4" src="https://user-images.githubusercontent.com/6199781/98874572-cda9cf00-247a-11eb-81df-612d061a5b94.png">

I want to take this chance to thank @epezent and all the colaborators of this project for this amazing library. Sincerely, thank you all for your amazing work ❤️

**Comment by ozlb** [comment]
> Not sure if this one counts, but since I will be using this library to make an application for USB oscilloscopes, I've been doing some experiments where I plotted some (generated) signals into heatmaps in order to visually differentiate the random noise from the real signal. 

@marcizhu Does the sample increase the "weight" and time decrease it? It will be great to have a "light" version in the demo...

**Comment by marcizhu** [comment]
Indeed, every "hit" of the signal increases  the histogram value by 1. I also keep a variable with the maximum value on the histogram at every given moment so that the colors cover the maximum possible range, but using time would also be a solution if you want a "fade out" effect once the signal disappears.

I can post a snippet with the code tomorrow if you guys want :)

**Comment by marcizhu** [comment]
Indeed, every "hit" of the signal increases  the histogram value by 1. I also keep a variable with the maximum value on the histogram at every given moment so that the colors cover the maximum possible range, but using time would also be a solution if you want a "fade out" effect once the signal disappears.

I can post a snippet with the code tomorrow if you guys want :)

**Comment by epezent** [comment]
@marcizhu, that's cool! What's the performance like? I didn't really optimize heatmaps for high density display, but it'd be an easy fix if it's needed.

**Comment by epezent** [comment]
Also, @marcizhu, feel free to submit a PR adding a minimal example to the demo :)

**Comment by ozlb** [comment]
> Indeed, every "hit" of the signal increases the histogram value by 1. I also keep a variable with the maximum value on the histogram at every given moment so that the colors cover the maximum possible range, but using time would also be a solution if you want a "fade out" effect once the signal disappears.
> 
> I can post a snippet with the code tomorrow if you guys want :)

"Fade out" effect will help you not only when signals disappears but generally when changes it's nature and a plot persistence factor will help you to track signal modulations

**Comment by marcizhu** [comment]
> @marcizhu, that's cool! What's the performance like? I didn't really optimize heatmaps for high density display, but it'd be an easy fix if it's needed.

@epezent The performance is really great: the histogram has a size of 1024x256, and rendering it on a 1280x720 window (the screen is 2560x1600) on a MacBook Pro 2018 (so not the best GPU in the world) I get 60-80 frames consistently. If we take into account the fact that my code is just a quick & dirty demo and it is not optimized at all, I'd say the performance is more than acceptable :)

> Also, @marcizhu, feel free to submit a PR adding a minimal example to the demo :)

Great! I will be opening a PR in a few days. Right now I'm a bit busy with university and exams, but once time allows I will definitely open a PR :+1:

> "Fade out" effect will help you not only when signals disappears but generally when changes it's nature and a plot persistence factor will help you to track signal modulations

@ozlb That's totally true. I will try using time instead of just tracking the maximum value of the histogram and if the end result looks nice, I will add that to the demo instead :D

**Comment by ejyanezp** [comment]
Here is an application to display seismic data, thanks to the **implot** project.
![Screenshot from 2020-11-18 22-57-35](https://user-images.githubusercontent.com/7320357/99619452-7b742b00-29f1-11eb-98dc-c15f0aa41c12.png)

**Comment by ddovod** [comment]
My deep reinforcement learning playground, heavily using different kinds of plots. Big thanks to implot!
![rl_implot](https://user-images.githubusercontent.com/9406025/101241535-f5055c00-3707-11eb-93c6-438c9dc628ee.png)

**Comment by epezent** [comment]
A quick little toy I put together while working on some [DSP algorithms](https://github.com/epezent/etfe).

![filter_toy](https://user-images.githubusercontent.com/29577475/104275075-c9446700-5467-11eb-8d86-19c2688b1951.gif)

**Comment by elvissteinjr** [comment]
[Desktop+](https://github.com/elvissteinjr/DesktopPlus) is a tool to bring your desktop and more into SteamVR. A recent addition is the Performance Monitor. 
ImPlot is used here to render the frame time graphs.
![performance monitor](https://user-images.githubusercontent.com/12502199/104975436-e7c3d880-59fa-11eb-8c87-e4d037105138.jpg)
This is pretty basic as far as plotting goes, but I just wanted something to render neat looking graphs in Dear ImGui and ImPlot fit the bill nicely. Thanks for that.

**Comment by jvannugteren** [comment]
I'm working on a GUI for a magnetic field calculation package called RAT. It uses GPU acceleration for magnetic field calculation and is therefore very fast compared to other codes. ImGui is used for the graphical user interface and ImPlot for all the 2D plotting. Although I'm not really using the real time capabilities of ImPlot, I think the library is great for my application as well.

![image](https://user-images.githubusercontent.com/78733220/110307861-b41b2d00-7fff-11eb-9905-71b932af9896.png)

**Reply by epezent** [reply]
That's awesome!

**Reply by epezent** [reply]
I should add that there is currently some work going into heatmap, and you should see a nice performance boost in the near future.

**Reply by jvannugteren** [reply]
Great. I indeed noticed a drop in the framerate when using too many pixels in the heatmaps.

**Comment by DrColossus** [comment]
I'm using ImPlot in a hobby project to analyze cycling workouts (power/heart rate/speed/GPS/etc.) 

![activity_viewer](https://user-images.githubusercontent.com/7056919/112457235-d8933b00-8d5b-11eb-8233-231757c28dbf.png)

I'm also downloading OSM tiles on the fly and use ImPlot to display maps ;)

![implot_map](https://user-images.githubusercontent.com/7056919/112457382-feb8db00-8d5b-11eb-8fd0-897a9f116c00.gif)

**Reply by epezent** [reply]
This is super cool! How exactly are you doing this? I assume you are using one or multiple calls to `PlotImage`. I would love to see the code if it is available somewhere!

**Reply by DrColossus** [reply]
Yes, exactly. The map uses Mercator projection normalized to [0, 1] and I use the tile resolution, ```GetPlotLimits``` and ```GetPlotSize``` to figure out which OSM tiles should be visible and then just issue a couple of ```PlotImage``` calls to display those tiles.

I'm planning to make it available at some point, but I haven't found the time yet :)

Here's how it works: OSM tiles use a simple x/y-coordinate and zoom level convention and there are some open tile servers available that you are free to use (with some restrictions like don't use more than 2 threads to download stuff). Each tile URL then consists of the base server URL concatenated with the tile coordinate and file extension (```/zoom/x/y.png```), i.e. the whole world in zoom level 0 can be found under https://a.tile.openstreetmap.org/0/0/0.png which I download using libcurl.
I do then cache the downloaded tiles on disk in a folder hierarchy that mirrors these ```url/zoom/x/y.png``` paths to avoid too many downloads.

Most of the logic happens when building a list of tiles from a coordinate bounding box.
So in pseudo-code, the logic is as follows:

    // We hold a map for all tiles with the
    // coordinate as search key
    map<coord, tile> cache
    vector<tile>     visible_tiles

    vector<tile> get_region(coord_min, coord_max)
    {
        // Just return the previous result if nothing changed
        if (coord_min,coord_max == previous_min,previous_max &&
            no_new_downloaded_tiles)
            return visible_tiles
        
        // The region changed, we rebuild the list
        // of visible tiles.
        visible_tiles.clear()
        
        for (coord in coord_min,coord_max)
        {
            tile = get_tile(coord)
            visible_tiles.push_back(tile)
        }
        
        return visible_tiles
    }

    tile get_tile(coord)
    {
        // The tile might already be in memory
        if (cache[coord] exists)
            return cache[coord]
            
        // We try to load it from disk or download it
        // if it is accessed for the very first time.
        load_tile_asynchronous(coord)
        
        // We still insert a lower res tile as fallback (it spans
        // a larger region, though). It will be replaced with the
        // full res tile once that becomes available.
        cache[coord] = find_lower_res(coord)
    }

    void load_tile_asynchronous(coord)
    {
        // This lambda function is pushed to a job queue
        // so that the filesystem access and/or download
        // does not block our UI code.
        lambda = [coord]()
        {
            filename = filename_from_coord(coord)
            path     = disk_cache_folder + filename
            
            if (!exists(path)
            {
                image = download_image(coord)
                write_to_disk(image, path)
                tile = tile_from_image(image)
                cache[coord] = tile
            }
            else
            {
                tile = load_from_file(path)
                cache[coord] = tile
            }
        }
        
        job_queue.push_back(lambda)
    }

There are some other shenanigans going on (like throwing away old tiles, sorting requested tiles by their distance to the plot center so they are loaded first, and so on) that were surprisingly simple and fun to implement.
This whole map rendering thing was basically a reason to learn a bit about web requests and ImPlot made it way easier by having all the input stuff just work, thank you for all the effort you put into this!

**Reply by epezent** [reply]
That's fantastic! I can't wait to play with it when you decide to make it available. It's worth noting that I'm currently working a [separate repository](https://github.com/epezent/implot_demos) to host more sophisticated demos of ImPlot. I think if you or I could add a minimal example of your maps to this, it would be a hit! Let me know what you think 😄

**Comment by epezent** [comment]
I've been playing with @marcizhu's awesome GPU accelerated heatmaps which have been merged into the `backends` branch (please, give it a try!). Here's a naïve Mandelbrot implementation, computed entirely on the CPU. Multithreading and AVX instructions keep the computation fast enough for smooth interaction, and the GPU accelerated heatmap renders the image in no time! 

![image](https://user-images.githubusercontent.com/29577475/126044244-8fcd3fda-f943-408f-babf-8dc6a42c5d60.png)

![image](https://user-images.githubusercontent.com/29577475/126044257-22487bdb-5bb4-4a91-b533-6fa49fc00b85.png)

![image](https://user-images.githubusercontent.com/29577475/126044302-a23d53c7-ea3e-4415-b323-68abba2c2aad.png)

**Reply by marcizhu** [reply]
Man that's awesome! Looks pretty nice, and the ~90 FPS seems really good to me!! I made a Mandelbrot viewer a while back using OpenMP for parallelism and SDL for rendering and I managed like 3 or 4 FPS using CPU-only rendering. Really hard to believe how fast this is. Good job!

By the way, will this be open sourced? I'd love to play with this by myself if that's possible 😄

**Reply by epezent** [reply]
Yes I'll put it in implot demos repo soon!

**Reply by epezent** [reply]
I got the actual Mandelbrot code from [here](https://github.com/skeeto/mandel-simd), and your reply just made me just realize what the ["strange" pragma](https://github.com/skeeto/mandel-simd/blob/1e9cf34e93cfaa60fba487e722dec2cc33e89ca3/mandel_avx.c#L16) I commented out was -- an OpenMP parallel for loop! I added this back with the right compiler flag, removed my ad-hoc threadpool, and now see 200 FPS for a 768x768 render!

**Comment by 0lru** [comment]
@see [https://github.com/0lru/p3ui](https://github.com/0lru/p3ui)
(PS: Object-Oriented Wrapper for ImPlot: [Header](https://github.com/0lru/p3ui/blob/main/cpp/p3/source/p3/Plot.h) [Source](https://github.com/0lru/p3ui/blob/main/cpp/p3/source/p3/Plot.cpp), wrapped to Python [right here](https://github.com/0lru/p3ui/blob/main/cpp/p3_python/source/p3/python/Plot.cpp))

![system](https://user-images.githubusercontent.com/61606260/126709034-1bdc80b4-d8fc-4de6-8b2e-6448a8faa4e4.png)

**Comment by epezent** [comment]
I've updated the [implot_demos](https://github.com/epezent/implot_demos) repo with several new demos that folks may enjoy. Note the repo is in need of cleaning up. You may experience a few issues with building.

**ImStocks**
- downloads historical data from Yahoo Finance and renders OHLC plot that is slightly more sophisticated than the one found in `implot_demo.cpp`

![stocks](https://user-images.githubusercontent.com/29577475/127591096-89e75736-207a-429b-80aa-ab3e4a90963f.png)

**ImMandel**
- see posts above

![mandel](https://user-images.githubusercontent.com/29577475/127591125-3bd76c89-c65a-4a17-b87e-a4f53a4cd766.png)


**ImMaps**
- my own implementation of what @DrColossus shared above

![maps](https://user-images.githubusercontent.com/29577475/127591138-0d7165e5-dc68-4069-9c5a-2f78eddaf93d.png)

**Reply by leeoniya** [reply]
nice, the vertical gridlines on OHLC are a bit odd. there are ones that land on month boundaries, but also fainter/mis-aligned ones that are a bit distracting.

**Reply by epezent** [reply]
Right, so that is the result of my time-axis implementation using months for the major ticks and weeks for the minor ticks at this zoom level. It felt logical to always restart the minor series on the first of the month. As a result, we sometimes get a minor tick right next to a major tick (e.g. 6/29 and 7/1 below).  I don't love it, but it doesn't bother me too much either, and we at least have a rational number of days between subdivisions. 

![image](https://user-images.githubusercontent.com/29577475/127599005-662c5cc6-2663-4b77-9961-1494d26c5f9c.png)

**Reply by epezent** [reply]
For comparison, uPlot (at least in this example), always has divisions in spans of weeks, but doesn't always show the first of the month (which I personally feel is useful, others may disagree). I would argue neither approach is perfect, and unfortunately it's just a function of the day/week/month levels being weird one to subdivided.

![image](https://user-images.githubusercontent.com/29577475/127600052-ae3a9fe2-f3d0-484f-b56a-1133385afca7.png)

**Reply by epezent** [reply]
Actually, I retract my comment as it doesn't seem uPlot is dividing by weeks in this example at all.

**Reply by epezent** [reply]
It took me way too long to realize you are omitting the weekends in this example. I see now that the ticks are in 10 day intervals. Doh!

**Reply by epezent** [reply]
![image](https://user-images.githubusercontent.com/29577475/127602262-7df4d3bf-5770-4f7e-b3a6-2d5d5e632ecb.png)

Better example of uPlot day ticks. Span is 4 days, which presents evenly spaced ticks doesn't always show a tick on the first of the month. Again, neither is wrong. Just a stylistic choice.

**Reply by leeoniya** [reply]
lol yeah, the ticks in that demo are ordinal and basically work the same way as if they were 0..N integers at the datapoints. then the tick density assesor picks some integer divisor. there is no knowledge of a time axis or calendar, besides formatting the timestamps at whatever datapoint ends up getting chosen. not great, but skipping weekends on discontinuous time scales is not something i'm willing to work through.

i do plan to improve it (see recent discussion in https://github.com/leeoniya/uPlot/issues/555), but it requires scanning all the data and assessing the Date() of each timestamp, which is okay for a demo with a few hundred datapoints, but maybe not thousands.

**Reply by leeoniya** [reply]
and i've agreed with you before, btw, about ticks on 1st of each month being a better choice, but obviously sacrifices a uniform grid in some cases.

**Reply by epezent** [reply]
As you mentioned, I have no idea how to support weekend removal using a continuous scale. Seems that even matplotlib and MATLAb are incapable of doing so without hackery by the user:
https://stackoverflow.com/questions/42155916/how-to-remove-weekends-in-matplotlib-candlestick-chart
https://www.mathworks.com/matlabcentral/answers/25643-excluding-weekend-gaps-from-financial-timeseries-plots

**Comment by dnedry2** [comment]


**Reply by epezent** [reply]
Incredible!

**Comment by bear24rw** [comment]
ImPlot histograms spotted in the Apple keynote today:

https://youtu.be/exM1uajp--A?t=1147

<img width="1484" alt="image" src="https://user-images.githubusercontent.com/778194/137782951-3566f312-b7e0-434c-827a-3be08ffb20dd.png">

**Reply by phicore** [reply]
How can you be sure about that?

**Reply by epezent** [reply]
I think @bear24rw is only referring to this plot. 

![image](https://user-images.githubusercontent.com/29577475/138113036-ef85fa62-f7cc-4f52-8af2-2f26579986ee.jpeg)

Definitely looks like ImGui. Can't be certain it's ImPlot though.

**Comment by epezent** [comment]
Sharing @CosminPerRam's incredible [sorting visualizer app](https://www.youtube.com/watch?v=UJPxN8VMNUY). The "Random Themes" section at the [4:23](https://www.youtube.com/watch?v=UJPxN8VMNUY&t=263s) mark is a fantastic demonstration of ImPlot's versatility!

https://github.com/CosminPerRam/StylishVisualizer

![image](https://user-images.githubusercontent.com/29577475/177901451-a726fc06-68c7-48b6-9a59-e4da5d4ca882.png)

**Reply by CosminPerRam** [reply]
Thanks for sharing it here. 
Yeah, ImPlot's versatiliy is great and it'll only become greater, I'm looking forward to update the app if more styles can be done!
The app couldn't be this fabulous without this good library, thanks!

**Comment by bear24rw** [comment]
Spotted in a SmarterEveryDay video:

https://youtu.be/dCXu8Ju_fdY?t=396

<img width="1056" alt="image" src="https://github.com/epezent/implot/assets/778194/2954a1f1-e596-4cd1-92c8-e2f4a4cf28d8">

**Reply by epezent** [reply]
This is awesome! I actually met Destin a long time ago in undergrad at the University of Alabama. He was sponsoring one of our senior design projects. Very cool dude.

**Comment by dnedry2** [comment]
RTLSDR Reciever with some options for live colormap changes / scaling. It can also record the incoming stream to a file

![brightness](https://github.com/epezent/implot/assets/31547542/8e035d81-ef78-4239-b46d-054403f6e938)
![colormap](https://github.com/epezent/implot/assets/31547542/9e9255d7-8a19-4036-a484-8db6a9ef037d)

**Reply by hinxx** [reply]
@dnedry2 cool waterfall widget! Any chance of sharing the source code for it?

**Reply by dnedry2** [reply]
@hinxx It's actually just an OpenGL texture that I render using PlotImage. All the coloring / scaling is done prior to the texture being passed to ImPlot
